### ISSUE TITLE
Allow lazy attachment of data sources

### ISSUE THREAD

**Issue by Morriz** [strongloop/loopback#2339]
Please decouple the attachment of data sources, so that models can be loaded synchronously. The need to unit test models is very real and this is a step in the right solution.

**Comment by gunjpan** [comment]
@Morriz : Hi, is this close to what you are looking for: https://github.com/strongloop/loopback/issues/2242?

**Comment by Morriz** [comment]
I have no clue anymore, I just wish that loopback was not so tightly coupled from the start. Is there a roadmap to refactor loopback so that it is decomposed into the unix style philosophy?

Ultimately I think it would be preferable to have:
- the data abstraction layer decoupled
- the server functionality decoupled (why that was tied so hard from the start I have no clue)
- the model layer decoupled (allowing unit testing)
- a cli environment to create jobs and tasks
- ???

**Comment by gunjpan** [comment]
@bajtos / @raymondfeng / @ritch : Could you please share your views re. ^ and their consideration in LB roadmap. Thanks.

**Comment by Morriz** [comment]
Sry if I come across a bit cynical, but I have been digging and trying to hack around all these impossibilities to come up with a way to create unit tests and simple task runs, but the system was never designed with that in mind, which makes me truly feel sad. I have been postponing a migration away from this mess for months now, but decided to ride with it, hoping it will become better in a pace that I can deal with. I guess it's time to roll our own. Unix-stylee...

However, I still hope you can offer insights into a roadmap out of this...

**Comment by superkhau** [comment]
@bajtos @raymondfeng ^

**Comment by bajtos** [comment]
@Morriz I am afraid I don't really understand what are you asking for.

> Please decouple the attachment of data sources, so that models can be loaded synchronously.

What do you mean by loading models synchronously? Could you please provide more details, perhaps a code sample showing what would you like to achieve?

> The need to unit test models is very real and this is a step in the right solution.

When you write "unit test models", are you referring to pure unit tests where the subject under test is completely isolated and all dependencies/integration points are mocked out?

I agree that pure unit tests may not be easy to achieve ATM. On the other hand, in my previous experience, mocking out the database/ORM layer can be tedious, and quite often it's better/easier to use in-memory database for tests. In LoopBack, we have `memory` connector that is perfect for writing fast tests, at least in theory. Have you tried it yet? What was the outcome?

Recently, we started to research ways how to decouple the implementation of the REST API from the ORM/data model. The idea is to introduce a concept of Controllers, which provide API independent of the data model, and defer data manipulation to Models (the same Models we have now).

Having said that, this is something you can already achieve now. Just create two models: one based on `Model` and providing the public API, the other based on `PersistedModel` and implementing data persistence.

> the data abstraction layer decoupled
> the server functionality decoupled (why that was tied so hard from the start I have no clue)
the model layer decoupled (allowing unit testing)

decoupled from what?

> a cli environment to create jobs and tasks

I am completely lost on this one. Could you please provide more details and some examples how would you use that? It seems to me like this item is unrelated to the rest of this discussion.

**Comment by Morriz** [comment]
I prefer a skype call or something, as I believe it will be waaaaaay faster to make you see what I mean. Are you open to that?

**Comment by Morriz** [comment]
Ok, let me try again. Please keep decoupling, modularity and unit testing in mind when you read my response.

And I don't want to come across as condescending here, but can I ask what the background is of the main coders of the stack? Monotlithic Java apps? It seems that they don't come from async programming at all. It would have been a good idea to have a crack aboard from the start and avoid all that tight coupling with time consuming async operations. Even now, that would still be a good idea imo.

> On 16 mei 2016, at 16:35, Miroslav Bajtoš <notifications@github.com> wrote:
> 
> @Morriz I am afraid I don't really understand what are you asking for.
> 
> Please decouple the attachment of data sources, so that models can be loaded synchronously.
> 
> What do you mean by loading models synchronously? Could you please provide more details, perhaps a code sample showing what would you like to achieve?
> 
Models should be just definitions, and not have any integration with resources that should be coupled lazily later on. So no datasources or any other stuff that defeats the purpose of synchronously loading the domain model. Why you ask me this question seems strange. The domain is just logic, and maybe some fancy enhancement to detect dirtiness. Look at Backbone. It has clean seperation of concerns.

Only when the domain layer needs to talk to async stuff, like backends, do you need to bootstrap that functionality. 

The whole point is to be able to have no async coupling with any outside stuff whatsoever. It's just bad design. Unit tests should be bloody fast, and not be hindered by ANY loading of unneeded code.

> The need to unit test models is very real and this is a step in the right solution.
> 
> When you write "unit test models", are you referring to pure unit tests where the subject under test is completely isolated and all dependencies/integration points are mocked out?
> 
Unit tests should not have dependencies on other functionality. Sure, you build upon third party code, but I am referring to your own work. When you need to stub you have bad design imo. Sure, sometimes you have old code and it hasn't been cut up nicely yet. 

Stubbing mostly makes sense when you want to do integration testing.
> I agree that pure unit tests may not be easy to achieve ATM. On the other hand, in my previous experience, mocking out the database/ORM layer can be tedious, and quite often it's better/easier to use in-memory database for tests. In LoopBack, we have memory connector that is perfect for writing fast tests, at least in theory. Have you tried it yet? What was the outcome?
> 
It still isn't what you'd want. Can we at least be on the same page with that? But yes, sometimes you have to make sure something is persisted, because (again) the code is not cleanly split up yet. And then the memory connector is making it less problematic.
> Recently, we started to research ways how to decouple the implementation of the REST API from the ORM/data model. The idea is to introduce a concept of Controllers, which provide API independent of the data model, and defer data manipulation to Models (the same Models we have now).
> 
> Having said that, this is something you can already achieve now. Just create two models: one based on Model and providing the public API, the other based on PersistedModel and implementing data persistence.
> 
Sounds like more obfuscating functionality. Just leave that up to the developers. The models should just be classes with static and prototyped functionality, extending from the dao. When 'async' stuff kicks in, fine, you have to wait for it to return.

And REST should have no part in this. The server should have no part in this either, so no express should run when there is no server context.
> the data abstraction layer decoupled
> the server functionality decoupled (why that was tied so hard from the start I have no clue)
> the model layer decoupled (allowing unit testing)
> 
> decoupled from what?
> 
Datasources, but actually anything hindring the straightforward testing of units of code.
> a cli environment to create jobs and tasks
> 
> I am completely lost on this one. Could you please provide more details and some examples how would you use that? It seems to me like this item is unrelated to the rest of this discussion.
> 
Again, I don't know why I have to explain myself here. I do hope that you understand the need to run tasks (scheduled, triggered etc).
It needs to be possible to run task, by having other entry points than just 'server'. I have to be able to create a folder 'jobs' with scripts that only require the models they need to do some tasks, persist the outcome, and quit. 
This is nowhere near possible right now. I spent days trying to decouple the monstrous spaghetti that boots everything, even the bloody boot scripts (which I had to create early exits in based on a flag). I have still not succeeded :|

I see the biggest refactor job in the context of loading models, and decoupling of their dependencies. IMHO your choices to have exotic loading mechanisms that deviate from javascript best practices are probably with good intentions: to offer convention over configuration, but it has become very messy and nearly impossible to work with. Integration hooks would have made it less painful, but I can'd find them.
Now you load models, do dependency loading, attach datasources (that have to connect directly, instead of lazily, just because you feel it important to decide for the developer to test wether it works immediately), then move the models to a previously bootstrapped (`boot()`) app.models. Etc etc. Plain ugly and obfuscation to the max.

What you would ultimately want is just to `import` your models, and that's it. Any functionality on the model that is async will lazily load whatever dependencies is has. So no more bootstrapping of unwanted functionality. And provide the loopback functionality as decoupled units ala unix philosophy.

I would love to architect a makeover of loopback, but I just have no time. But my advice: unix philosophy. Stop making 'automagic' monoliths.

**Comment by bajtos** [comment]
@Morriz thank you for the long comment, I think I am starting to understand what you are asking for.

Before I respond to your comment, I am kindly asking you to refrain from making comments about personalities of people involved here and keep the discussion focused on the issues/features. 

Few quotes showing what I mean:
> And I don't want to come across as condescending here, but can I ask what the background is of the main coders of the stack? Monotlithic Java apps? It seems that they don't come from async programming at all.

> Again, I don't know why I have to explain myself here. I do hope that you understand the need to run tasks (scheduled, triggered etc).

---

In general, you don't have to use loopback-boot and it's conventions in your LoopBack apps. We are very careful to separate concerns and keep loopback-boot just a (relatively) thin wrapper delegating as much to LoopBack core as possible.

For example, you can create your models this way:

```js
var app = loopback();
// ...

var MyModel = app.registry.createModel('MyModel', { /* properties */ }, { /* settings */ });
// or using loopback-boot's JSON structure:
var MyModel = app.registry.createModel({
  name: 'MyModel',
  properties: { /* properties */ },
  /* settings are at the top level */
});

// Now you can add custom methods
MyModel.greet = function(cb) {
  cb(null, 'hello world');
};

MyModel.remoteMethod('greet', {
  // etc.
});
```

Similarly,  there is API for building data-sources and attaching models to datasources:

```js
app.dataSource('db', { connector: 'memory' });
app.model(MyModel, {
  dataSource: 'db',
  public: false // set to true to expose via REST
});
```

If the conventions of loopback-boot don't work for you, then I am encouraging you to define your own conventions and perhaps implement an alternative bootstrapper.

> Models should be just definitions, and not have any integration with resources that should be coupled lazily later on. So no datasources or any other stuff that defeats the purpose of synchronously loading the domain model.  Why you ask me this question seems strange. The domain is just logic, and maybe some fancy enhancement to detect dirtiness. Look at Backbone. It has clean seperation of concerns. 

I am not familiar with Backbone models, but I believe LoopBack models actually satisfy your requirement *models should be just definitions and not have any integrations with resources*. When you create a model via `app.registry.createModel`, it is not integrated with any resources. You have to attach it to a datasource to give it async methods.

Having wrote that, I admit that we don't have lazy attach that would be triggered by the first call of a methods that requires it.

> It needs to be possible to run task, by having other entry points than just 'server'. I have to be able to create a folder 'jobs' with scripts that only require the models they need to do some tasks, persist the outcome, and quit.

I see, this makes sense. The project structure and conventions supported by loopback-boot were optimized for different scenarios. However, I think you can still implement a script that requires a subset of models only.

Let me explain: the list of models loaded and configured by loopback-boot is driven by entries in `server/model-config.json`. We look at what models are used by the app, and then define only these models (plus any parent/base models as needed). A naive first try would be to create a new subdirectory in your project (e.g. "tasks" living beside "common" and "server") and place different `server/model-config.json` there. Of course, that does not work very for more that one job/task script when each of them requires different subset of models.

Fortunately, loopback-boot allows callers to provide custom model configuration overriding configuration provided in the JSON file, see [source](https://github.com/strongloop/loopback-boot/blob/master/lib/compiler.js#L46-L47). I imagine your job/task script do something like this:

```js
var serverModels = require('../server/model-config.json');
var scriptModels = {};
for (var key in serverModels) {
  if (shouldLoadModel(key)) {
    scriptModels[key] = serverModels[key];
  }
}
  
var app = loopback();
boot(app, {
  appRootDir: path.resolve('..', 'server');
  models: scriptModels,
});
```

I admit it's not as pretty as calling `require('path/to/my-model.js')` from a job/task script.

**Comment by Morriz** [comment]
Wow, tnx for the grand reply!
And sry for my lack of netiquette, still working on it :|

I was under the assumption that it wasnt possible to load the models properly and configured with a backend in a synchronous fashion, but I had a breakthrough splitting up the code and now have it working as intended. I will add it to the demo-api loopback stack I published asap.

Might very well be the best thing to do imo, update your example stacks. And with more valuable use cases, as now they are mostly very minimal and pooc style

**Comment by bajtos** [comment]
Looks like all is good, I am closing this issue as resolved.
